From e8812b6d7b42cd514e57465c9d901a70ee52f8f4 Mon Sep 17 00:00:00 2001
From: Colin Watson <cjwatson@ubuntu.com>
Date: Fri, 28 Mar 2014 17:09:45 +0000
Subject: Remove limits on loop labels

There's no reason to impose any particular limit on loop labels, since
they just represent a single large partition.  Sector counts over 2^32
are needed for large RAID arrays.

Bug-Debian: http://bugs.debian.org/592679
Bug-Ubuntu: https://bugs.launchpad.net/bugs/543838
Forwarded: http://lists.alioth.debian.org/pipermail/parted-devel/2010-August/003725.html
Last-Update: 2010-04-09

Patch-Name: loop-limits.patch
---
 libparted/labels/pt-limit.c     | 8 ++++----
 libparted/labels/pt-limit.gperf | 3 ++-
 tests/t9021-maxima.sh           | 2 +-
 3 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/libparted/labels/pt-limit.c b/libparted/labels/pt-limit.c
index bda2ae4..1888f36 100644
--- a/libparted/labels/pt-limit.c
+++ b/libparted/labels/pt-limit.c
@@ -120,8 +120,8 @@ pt_limit_lookup (register const char *str, register unsigned int len)
 #line 18 "pt-limit.gperf"
       {"sun",128ULL*UINT32_MAX,UINT32_MAX},
       {""}, {""}, {""}, {""},
-#line 22 "pt-limit.gperf"
-      {"loop",UINT32_MAX,UINT32_MAX},
+#line 23 "pt-limit.gperf"
+      {"loop",UINT64_MAX,UINT64_MAX},
       {""}, {""}, {""}, {""},
 #line 10 "pt-limit.gperf"
       {"gpt",UINT64_MAX,UINT64_MAX},
@@ -132,7 +132,7 @@ pt_limit_lookup (register const char *str, register unsigned int len)
 #line 8 "pt-limit.gperf"
       {"dasd",UINT32_MAX,UINT32_MAX},
       {""}, {""}, {""}, {""},
-#line 26 "pt-limit.gperf"
+#line 27 "pt-limit.gperf"
       {"amiga",UINT32_MAX,UINT32_MAX},
       {""}, {""}, {""}, {""},
 #line 9 "pt-limit.gperf"
@@ -144,7 +144,7 @@ pt_limit_lookup (register const char *str, register unsigned int len)
 #line 20 "pt-limit.gperf"
       {"bsd",UINT32_MAX,UINT32_MAX},
       {""}, {""}, {""}, {""},
-#line 23 "pt-limit.gperf"
+#line 24 "pt-limit.gperf"
       {"pc98",UINT32_MAX,UINT32_MAX}
     };
 
diff --git a/libparted/labels/pt-limit.gperf b/libparted/labels/pt-limit.gperf
index f834647..3d764ae 100644
--- a/libparted/labels/pt-limit.gperf
+++ b/libparted/labels/pt-limit.gperf
@@ -19,7 +19,8 @@ sun,128ULL*UINT32_MAX,UINT32_MAX
 #
 bsd,UINT32_MAX,UINT32_MAX
 # aix,UINT32_MAX,UINT32_MAX
-loop,UINT32_MAX,UINT32_MAX
+# In reality, loop labels have no particular limit.
+loop,UINT64_MAX,UINT64_MAX
 pc98,UINT32_MAX,UINT32_MAX
 #
 # FIXME: not verified.  looks like these are cylinder aligned, too
diff --git a/tests/t9021-maxima.sh b/tests/t9021-maxima.sh
index eb44eea..0f16297 100644
--- a/tests/t9021-maxima.sh
+++ b/tests/t9021-maxima.sh
@@ -42,7 +42,7 @@ for t in msdos gpt dvh sun mac bsd amiga loop pc98; do
     max_start=4294967295
     max_len=4294967295
     case $t in
-	gpt) max_start=18446744073709551615; max_len=$max_start;;
+	gpt|loop) max_start=18446744073709551615; max_len=$max_start;;
 	sun) max_start=549755813760;; # 128 * (2^32-1)
     esac
 
